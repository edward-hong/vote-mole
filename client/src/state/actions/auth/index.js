import { FETCH_USER, FETCH_USER_FULFILLED } from '../../../constants'

export const fetchUser = () => ({ type: FETCH_USER })

export const fetchUserFulfilled = payload => ({
	type: FETCH_USER_FULFILLED,
	payload,
})
