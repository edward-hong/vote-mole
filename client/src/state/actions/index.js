export * from './auth'
export * from './polls'
