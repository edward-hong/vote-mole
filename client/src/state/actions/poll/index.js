import {
	POLL_INFO_TYPE,
	POLL_INFO_RECEIVED_TYPE,
	POLL_CLEAR_TYPE,
} from '../../../constants'

// Fetch poll
export const fetchPoll = payload => ({
	type: POLL_INFO_TYPE,
	payload,
})

// Dispatch received poll data to reducer
export const dispatchPollReceived = payload => ({
	type: POLL_INFO_RECEIVED_TYPE,
	payload,
})

// Action for clearing poll data from state
export const clearPoll = () => ({ type: POLL_CLEAR_TYPE })
