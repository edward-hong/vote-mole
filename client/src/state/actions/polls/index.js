import {
	POLL_ALL_TYPE,
	POLL_ALL_RECEIVED_TYPE,
	POLL_CLEAR_TYPE,
	POLL_USER_TYPE,
	POLL_USER_RECEIVED_TYPE,
} from '../../../constants'

// Fetch all polls
export const fetchAllPolls = payload => ({
	type: POLL_ALL_TYPE,
	payload,
})

// Dispatch received all polls to reducer
export const dispatchAllPollsReceived = payload => ({
	type: POLL_ALL_RECEIVED_TYPE,
	payload,
})

// Fetch user polls
export const fetchUserPolls = payload => ({
	type: POLL_USER_TYPE,
	payload,
})

// Dispatch received user polls to reducer
export const dispatchUserPollsReceived = payload => ({
	type: POLL_USER_RECEIVED_TYPE,
	payload,
})

// Clear polls state
export const clearPolls = () => ({ type: POLL_CLEAR_TYPE })
