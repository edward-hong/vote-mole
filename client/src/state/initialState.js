const initialState = {
	auth: null,
	poll: null,
}

export default initialState
