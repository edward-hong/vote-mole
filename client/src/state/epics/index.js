import { combineEpics } from 'redux-observable'
import { values } from 'ramda'

import * as authEpics from './auth'
import * as pollEpics from './polls'

const rootEpic = combineEpics(...values(authEpics), ...values(pollEpics))

export default rootEpic
