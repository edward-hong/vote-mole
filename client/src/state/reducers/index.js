import { combineReducers } from 'redux'
import { reducer as form } from 'redux-form'

import auth from './auth'
import polls from './polls'

const rootReducer = combineReducers({ auth, form, polls })

export default rootReducer
