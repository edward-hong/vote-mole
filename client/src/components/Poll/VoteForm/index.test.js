import React from 'react'
import { fireEvent } from '@testing-library/react'

import VoteForm from '.'
import renderWithRedux from '../../../utils/test/renderWithRedux'
import { votePoll } from '../../../state/actions'

jest.mock('../../../state/actions', () => ({
	votePoll: jest.fn(),
}))
global.open = jest.fn()

votePoll.mockImplementation(() => ({ type: 'test' }))

const poll = {
	_id: '5cdaf28cee882077685935e3',
	userId: '5cd19682498a98c6962f34c1',
	pollQuestion: 'PHP, Python or Ruby?',
	pollOptions: [
		{
			votes: 3,
			_id: '5cdaf28cee882077685935e6',
			option: 'PHP',
		},
		{
			votes: 8,
			_id: '5cdaf28cee882077685935e5',
			option: 'Python',
		},
		{
			votes: 6,
			_id: '5cdaf28cee882077685935e4',
			option: 'Ruby',
		},
	],
}

describe('VoteForm', () => {
	it('renders Loader when there is no poll', () => {
		const { container } = renderWithRedux(<VoteForm />, {
			initialState: { form: { vote: {} }, auth: false, poll: undefined },
		})

		expect(container.firstChild).toMatchSnapshot()
	})

	it('renders correctly when user is not logged in', () => {
		const { container } = renderWithRedux(<VoteForm />, {
			initialState: { form: { vote: {} }, auth: false, poll },
		})

		expect(container.firstChild).toMatchSnapshot()
	})

	it('shows custom option error when user selects custom option but have not filled out field', () => {
		const auth = {
			_id: '5cd19682498a98c6962f34c1',
			profileID: '111881712224957728911',
		}
		const vote = {
			registeredFields: {
				selection: {
					name: 'selection',
					type: 'Field',
					count: 1,
				},
				customSelection: {
					name: 'customSelection',
					type: 'Field',
					count: 1,
				},
			},
			fields: {
				selection: {
					visited: true,
					touched: true,
				},
				customSelection: {
					visited: true,
					touched: true,
				},
			},
			values: {
				selection: "I'd like a custom option",
			},
			syncErrors: {
				customSelection: 'Please enter your custom selection',
			},
			anyTouched: true,
		}
		const { getByText } = renderWithRedux(<VoteForm />, {
			initialState: {
				form: { vote },
				auth,
				poll,
			},
		})
		const errorMessage = getByText('Please enter your custom selection')

		expect(errorMessage).toBeInTheDocument()
	})

	it('calls submit when submitting form', () => {
		const vote = {
			registeredFields: {
				selection: {
					name: 'selection',
					type: 'Field',
					count: 1,
				},
			},
			fields: {
				selection: {
					visited: true,
					active: true,
				},
			},
			active: 'selection',
			values: {
				selection: 'PHP',
			},
		}
		const { getByTestId } = renderWithRedux(<VoteForm />, {
			initialState: {
				form: { vote },
				auth: false,
				poll,
			},
		})
		const form = getByTestId('vote-form')

		fireEvent.submit(form)

		expect(votePoll).toHaveBeenCalled()
	})

	it('sends user to Twitter when tweet button is clicked', () => {
		const auth = {
			_id: '5cd19682498a98c6962f34c1',
			profileID: '111881712224957728911',
		}
		const { getByText } = renderWithRedux(<VoteForm />, {
			initialState: {
				form: { vote: {} },
				auth,
				poll,
			},
		})
		const tweetButton = getByText('Tweet')
		fireEvent.click(tweetButton)

		expect(global.open).toHaveBeenCalled()
	})
})
