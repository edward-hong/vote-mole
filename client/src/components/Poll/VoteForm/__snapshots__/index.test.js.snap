// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoteForm renders Loader when there is no poll 1`] = `
.c0 {
  position: relative;
  margin-top: 30px;
}

.c1 {
  margin: -24px 0 0 -24px;
  height: 49px;
  width: 49px;
  -webkit-animation: cssload-rotate 1150ms linear infinite;
  animation: cssload-rotate 1150ms linear infinite;
}

.c1,
.c1:before,
.c1:after {
  position: absolute;
  top: 50%;
  left: 50%;
  border: 1px solid #ccc;
  border-left-color: black;
  border-radius: 974px;
}

.c1:before {
  content: '';
  margin: -22px 0 0 -22px;
  height: 43px;
  width: 43px;
  -webkit-animation: cssload-rotate 1150ms linear infinite;
  animation: cssload-rotate 1150ms linear infinite;
}

.c1:after {
  content: '';
  margin: -28px 0 0 -28px;
  height: 55px;
  width: 55px;
  -webkit-animation: cssload-rotate 2300ms linear infinite;
  animation: cssload-rotate 2300ms linear infinite;
}

<div
  class="col-12"
>
  <div
    class="c0"
  >
    <div
      class="c1"
    />
  </div>
</div>
`;

exports[`VoteForm renders correctly when user is not logged in 1`] = `
.c0 {
  font-size: 1.5rem;
  margin-bottom: 1rem;
}

.c1 {
  display: block;
}

@media (max-width:575px) {
  .c0 {
    font-size: 2.5rem;
    margin-bottom: 1.5rem;
    text-align: center;
  }
}

<div
  class="col-12 col-sm-4"
>
  <h1
    class="c0"
  >
    PHP, Python or Ruby?
  </h1>
  <form
    class=""
    data-testid="vote-form"
    id="vote"
  >
    <div
      class="form-group"
    >
      <label
        class=""
        for="selection"
      >
        I want to vote for...
      </label>
      <select
        class="form-control"
        name="selection"
      >
        <option
          disabled=""
        />
        <option>
          PHP
        </option>
        <option>
          Python
        </option>
        <option>
          Ruby
        </option>
      </select>
    </div>
    <div
      class="c1 btn-group-vertical"
      role="group"
    >
      <button
        class="btn btn-outline-success"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
